paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java

```diff
@@ -162,16 +172,45 @@ public RecordReader<InternalRow> createReader(DataSplit split) throws IOExceptio
                                         Projection.of(dataProjection)
                                                 .project(dataSchema.logicalRowType());
 
-                                return new BulkFormatMapping(
-                                        indexCastMapping.getIndexMapping(),
-                                        indexCastMapping.getCastMapping(),
-                                        partitionPair,
-                                        formatDiscover
-                                                .discover(formatIdentifier)
-                                                .createReaderFactory(
-                                                        projectedRowType, dataFilters));
+                                return Triple.of(
+                                        dataSchema,
+                                        dataFilters,
+                                        new BulkFormatMapping(
+                                                indexCastMapping.getIndexMapping(),
+                                                indexCastMapping.getCastMapping(),
+                                                partitionPair,
+                                                formatDiscover
+                                                        .discover(formatIdentifier)
+                                                        .createReaderFactory(
+                                                                projectedRowType, dataFilters)));
                             });
 
+            TableSchema dataSchema = formatMappingTriple.f0;
+            List<Predicate> dataFilter = formatMappingTriple.f1;
+            BulkFormatMapping bulkFormatMapping = formatMappingTriple.f2;
+            if (dataFilter != null && !dataFilter.isEmpty()) {
+                List<String> indexFiles =
+                        file.extraFiles().stream()
+                                .filter(
+                                        name ->
+                                                name.startsWith(
+                                                        DataFilePathFactory.INDEX_PATH_PREFIX))
+                                .collect(Collectors.toList());
+                if (fileIndexReadEnabled && !indexFiles.isEmpty()) {
```

Comment:
Here should just be a simple wrapper `RecordReader`. Just like `ApplyDeletionVectorReader`.